File: packages/web/src/api/auth.api.ts
import { createApi } from '@reduxjs/toolkit/query/react'

import baseQueryWithReauth from './baseQueryWithReauth'

import { SignUpDto } from '~types/auth/sign-up.dto'
import { IUser } from '~types/user/user.type'
import { VerifyEmailDto, VerifyEmailResponse } from '~types/auth/verify-email.type'
import { LoginResponse } from '~types/auth/login-response.type'
import { LoginDto } from '~types/auth/login.dto'
import { ForgetPasswordRequest } from '~types/auth/forget-password-verify.type'
import { ChangePasswordRequest } from '~types/auth/change-password.type'

export const api = createApi({
  reducerPath: 'authApi',
  baseQuery: baseQueryWithReauth,
  endpoints: builder => ({
    register: builder.mutation<IUser, SignUpDto>({
      query: body => ({
        url: 'auth/register',
        method: 'POST',
        body,
      }),
    }),
    verifyEmail: builder.mutation<VerifyEmailResponse, VerifyEmailDto>({
      query: body => ({
        url: 'auth/verify-email',
        method: 'POST',
        body,
      }),
    }),
    login: builder.mutation<LoginResponse, LoginDto>({
      query: body => ({
        url: 'auth/login',
        method: 'POST',
        body,
      }),
    }),
    logout: builder.mutation<void, void>({
      query: () => ({
        url: 'auth/logout',
        method: 'POST',
      }),
    }),
    passwordVerificationCode: builder.mutation<{ message: string }, ForgetPasswordRequest>({
      query: body => ({
        url: 'auth/password-verification-code',
        method: 'PATCH',
        body,
      }),
    }),
    changePassword: builder.mutation<{ message: string }, ChangePasswordRequest>({
      query: body => ({
        url: 'auth/change-password',
        method: 'PATCH',
        body,
      }),
    }),
    googleAuth: builder.query<void, void>({
      query: () => ({
        url: 'auth/google-login',
        method: 'GET',
      }),
    }),
    githubAuth: builder.query<void, void>({
      query: () => ({
        url: 'auth/github-login',
        method: 'GET',
      }),
    }),
  }),
})

export const {
  useRegisterMutation,
  useVerifyEmailMutation,
  useLoginMutation,
  usePasswordVerificationCodeMutation,
  useChangePasswordMutation,
  useLogoutMutation,
  useLazyGoogleAuthQuery,
  useLazyGithubAuthQuery,
} = api

File: packages/mobile/app/src/main/java/com/devhub/devhubapp/api/AuthAPI.kt
package com.devhub.devhubapp.api

import com.devhub.devhubapp.dataClasses.ChangePasswordRequest
import com.devhub.devhubapp.dataClasses.LoginRequest
import com.devhub.devhubapp.dataClasses.LoginResponse
import com.devhub.devhubapp.dataClasses.PasswordVerificationRequest
import com.devhub.devhubapp.dataClasses.RegistrationRequest
import com.devhub.devhubapp.dataClasses.RegistrationResponse
import com.devhub.devhubapp.dataClasses.TokenResponse
import com.devhub.devhubapp.dataClasses.VerifyEmailRequest
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.PATCH
import retrofit2.http.POST

interface AuthAPI {

    @POST("auth/register")
    fun register(@Body registrationRequest: RegistrationRequest): Call<RegistrationResponse>

    @POST("auth/verify-email")
    fun verifyEmail(@Body verifyEmailRequest: VerifyEmailRequest): Call<ResponseBody>

    @POST("auth/login")
    fun login(@Body loginRequest: LoginRequest): Call<LoginResponse>

    @PATCH("auth/password-verification-code")
    fun passwordVerificationCode(@Body passwordVerificationRequest: PasswordVerificationRequest): Call<ResponseBody>

    @PATCH("auth/change-password")
    fun changePassword(@Body changePassword: ChangePasswordRequest): Call<ResponseBody>

    @POST("auth/refresh")
    fun refreshToken(
        @Header("X-Refresh-Token") refreshToken: String
    ): Call<TokenResponse>

    @GET("auth/testinfo")
    fun testinfo(
        @Header("Authorization") accessToken: String
    ): Call<ResponseBody>

}